src/pint/models/binary_piecewise.py

```diff
@@ -0,0 +1,325 @@
+            prefixParameter(
+                name="PLB_{0}".format(i),
+                units="MJD",
+                unit_template=lambda x: "MJD",
+                description="Beginning of paramX interval",
+                description_template=lambda x: "Beginning of paramX interval",
+                parameter_type="MJD",
+                time_scale="utc",
+                value=group_start_mjd,
+            )
+        )
+        self.add_param(
+            prefixParameter(
+                name="PUB_{0}".format(i),
+                units="MJD",
+                unit_template=lambda x: "MJD",
+                description="End of paramX interval",
+                description_template=lambda x: "End of paramX interval",
+                parameter_type="MJD",
+                time_scale="utc",
+                value=group_end_mjd,
+            )
+        )
+        # print("going to binary_piecewise setup")
+        # self.setup()
+        # self.validate()
+
+    def update_binary_object(self, toas=None, acc_delay=None):
+        super().update_binary_object(toas=toas, acc_delay=acc_delay)
+
+    def d_binary_delay_d_xxxx(self, toas, param, acc_delay):
+        """Return the binary model delay derivtives."""
+        # print(f"hello from binary piecewise parameter derivative, using toas: {toas}")
+        delay = super().d_binary_delay_d_xxxx(
+            toas=toas, param=param, acc_delay=acc_delay
+        )
+        return delay
+
+    def remove_range(self, index):
+        """Removes all DMX parameters associated with a given index/list of indices.
+
+        Parameters
+        ----------
+
+        index : float, int, list, np.ndarray
+            Number or list/array of numbers corresponding to DMX indices to be removed from model.
+        """
+
+        if (
+            isinstance(index, int)
+            or isinstance(index, float)
+            or isinstance(index, np.int64)
+        ):
+            indices = [index]
+        elif not isinstance(index, list) or not isinstance(index, np.ndarray):
+            raise TypeError(
+                f"index must be a float, int, list, or array - not {type(index)}"
+            )
+        for index in indices:
+            index_rf = f"{int(index):04d}"
+            for prefix in ["T0X_", "A1X_", "PLB_", "PUB_"]:
+                self.remove_param(prefix + index_rf)
+        self.setup()
+        self.validate()
+
+    def add_piecewise_param(self, param, param_unit, paramx, j):
+        if int(j) in self.get_prefix_mapping_component("X_"):
+            raise ValueError(
+                "Index '%s' is already in use in this model. Please choose another." % j
+            )
+        if j is None:
+            dct = self.get_prefix_mapping_component(param + "X_")
+            j = np.max(list(dct.keys())) + 1
+        i = f"{int(j):04d}"
+        if param is "A1":
+            self.add_param(
+                prefixParameter(
+                    name=param + "X_{0}".format(i),
```

Comment:
Nicer to use `f"X_{i}"`

Reply:
Changed string formatting as suggested (here and in other places).